### PR TITLE
Scripting/AP_Arming: add pre-arm check that scripting has initialised successfull

### PR DESCRIPTION
**Pull request by rmackay9**
This PR resolves issue https://github.com/ArduPilot/ardupilot/issues/12798 by adding a pre-arm check to alert the user if scripting has failed to initialise because it has run out of memory.  In my tests a CubeBlack (and thus probably many other boards) are very close to running out of memory when using Copter in the default configuration so it is fairly likely that users will hit this issue and so we should provide protection and a decent warning.

This also removes the automatic setting of SCR_ENABLE to zero which we've recently agreed is not a good pattern because users may inadvertantly save/re-load parameter and permanently disable scripting.

This has been tested on a real board and a screen shot of the pre-arm check message is shown below.
image

This PR also makes an unrelated change to print an initialisation failure message from the scripting library itself instead of duplicating it in each vehicle.

### CODE REVIEW

File: libraries/AP_Scripting/AP_Scripting.h [hunk 1]
```diff
@@ -28,9 +28,10 @@ class AP_Scripting
     AP_Scripting(const AP_Scripting &other) = delete;
     AP_Scripting &operator=(const AP_Scripting&) = delete;
 
-    bool init(void);
+    void init(void);
+    bool init_failed(void) const { return _init_failed; }
 
-    bool enabled(void) const { return _enable != 0; };
+    bool enabled(void) const { return (_enable != 0) && !init_failed(); };
```
Comment (WickedShell) [inline libraries/AP_Scripting/AP_Scripting.h]: Do we really want to hook this here? In a lot of ways you want to be able to distinguish between "does the user want this to run right now" and "we wanted to run, but can't". As far as I can tell this isn't actually needed for this PR to do it's desired job.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 2]
```diff
@@ -705,6 +706,13 @@ bool AP_Arming::system_checks(bool report)
             check_failed(ARMING_CHECK_SYSTEM, report, "Terrain out of memory");
             return false;
         }
+#endif
+#ifdef ENABLE_SCRIPTING
+        const AP_Scripting *scripting = AP_Scripting::get_singleton();
+        if ((scripting != nullptr) && scripting->init_failed()) {
```
Comment (WickedShell) [inline libraries/AP_Arming/AP_Arming.cpp]: I'd recommend removing the changes to `enabled()` and then adding a check here if scripting was enabled. It's quite conceivable you might want to just not use scripting, and just set the enable flag to 0, which wouldn't require a reboot to stop scripts. (The other way would ofc). The main target is just to make it easier to get around this check if you don't need scripts without requiring a reboot.

File: libraries/AP_Scripting/AP_Scripting.h (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_Scripting/AP_Scripting.h]: I've reverted this change, thanks.

File: libraries/AP_Arming/AP_Arming.cpp (hunk 2 above)
Comment (rmackay9) [reply libraries/AP_Arming/AP_Arming.cpp]: As requested, I've added a check here that scripting was enabled, thanks

**Comment by rmackay9** [comment]
Merged after discussing with @WickedShell

**Comment by rmackay9** [comment]
this has been added to Copter-4.0.0-rc3

**Pull request merged**